Add the New Save Panel

Calisto-Mathias opened the pull request:
In this commit, I've added a new save panel accessible from the Menu Bar. Users can name and mark a query as a favorite while saving it. They can choose to save the query in the default directory or a user-selected directory. If the default directory option is unchecked, a BFilePanel opens to set the directory. The save button is disabled until a name is entered in the BTextControl. Currently, there's no option to change the default directory, but this may be added later.

nielx reviewed (changes requested):
Added some notes and suggestions. I will do a more proper review and testing, but this should give you some hints to get started with. I don't expect any fundamental comments, at this first pass the code looks more than decent.

Calisto-Mathias commented on src/kits/tracker/FindPanel.h:
```diff
@@ -119,6 +127,31 @@ struct MoreOptionsStruct {
+								SaveWindow(BWindow*);	
+	virtual						~SaveWindow();
+	virtual	bool				QuitRequested();
+private:
```
I think I fixed this while extracting the code into separate files. I made sure to leave a line between the protected, private, and public sections of the class.

Calisto-Mathias commented on src/kits/tracker/FindPanel.h:
```diff
@@ -119,6 +127,31 @@ struct MoreOptionsStruct {
+			SavePanel*			fSavePanel;
+};
+
+class SavePanel: public BView{
```
I had been following a general pattern observed in other code, where the Panel was always separated. This was mainly due to the separation of concerns necessitated by the many interactions in the Views. However, I realized that my Window is relatively small and contains simple interactions. Therefore, I decided to consolidate the code into SaveWindow.h and SaveWindow.cpp files , moving all the interaction logic into a single class and its implementation.

Thanks for the tip, Niels! I will note it down and keep it in mind for future implementations.

Calisto-Mathias commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -3447,9 +3553,120 @@ MostUsedNames::UpdateList()
 	fList.SortItems(MostUsedNames::CompareNames);
 }
 
+SaveWindow::SaveWindow(BWindow *parentWindow)
```
Yes, in my latest commit, I moved this into its own separate file.

Calisto-Mathias commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -3447,9 +3553,120 @@ MostUsedNames::UpdateList()
 
 	fList.SortItems(MostUsedNames::CompareNames);
 }
 
```
I think I fixed this in the latest commit as well.

Calisto-Mathias commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -420,11 +430,38 @@ FindWindow::SaveQueryAttributes(BNode* file, bool queryTemplate)
 FindWindow::SaveQueryAsAttributes(BNode* file, BEntry* entry,
 	bool queryTemplate, const BMessage* oldAttributes,
-	const BPoint* oldLocation)
+	const BPoint* oldLocation,bool includeInFavorites)
```
Noted. I fixed this. I also fixed other parts of the code where I hadn't given a space after appropriate delimiters.

Calisto-Mathias commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -45,6 +45,7 @@ All rights reserved.
 #include <strings.h>
 
 #include <Application.h>
+#include <MenuBar.h>
```
Fixed this in the newer commit. I also followed the same alphabetical-order guideline in the newer SaveWindow.h and SaveWindow.cpp files.

nielx commented:
The current code seems to end up in a linker error for me.

```
/DevData1/haiku-build> jam -q Tracker
warning: Invalid jamfile cache: Failed to read file info.
Starting build of type regular ... 
AddHaikuImagePackages: package gutenprint8 not available! 
Asked for bios_ia32 target boot platform 
Unknown path to handle adding to image 
Asked for pxe_ia32 target boot platform 
Unknown path to handle adding to image 
...patience...
...patience...
...found 13655 target(s)...
...updating 2 target(s)...
Link objects/haiku/x86_64/release/kits/tracker/libtracker.so 
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: objects/haiku/x86_64/release/kits/tracker/FindPanel.o: in function `BPrivate::FindWindow::MessageReceived(BMessage*)':
FindPanel.cpp:(.text+0x8814): undefined reference to `BPrivate::SaveWindow::SaveWindow(BWindow*)'
collect2: error: ld returned 1 exit status

gcc -fno-strict-aliasing -fno-delete-null-pointer-checks -fno-builtin-fork -fno-builtin-vfork -Xlinker --no-undefined -shared -Xlinker -soname="libtracker.so" -nostdlib -Xlinker --no-undefined -o "objects/haiku/x86_64/release/kits/tracker/libtracker.so"  "objects/haiku/x86_64/release/system/glue/arch/x86_64/crti.o" "/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/crtbeginS.o" "objects/haiku/x86_64/release/system/glue/init_term_dyn.o" "objects/haiku/x86_64/release/kits/tracker/AttributeStream.o" "objects/haiku/x86_64/release/kits/tracker/AttributesView.o" "objects/haiku/x86_64/release/kits/tracker/AutoMounterSettings.o" "objects/haiku/x86_64/release/kits/tracker/BackgroundImage.o" "objects/haiku/x86_64/release/kits/tracker/Bitmaps.o" "objects/haiku/x86_64/release/kits/tracker/ContainerWindow.o" "objects/haiku/x86_64/release/kits/tracker/CountView.o" "objects/haiku/x86_64/release/kits/tracker/DeskWindow.o" "objects/haiku/x86_64/release/kits/tracker/DesktopPoseView.o" "objects/haiku/x86_64/release/kits/tracker/DraggableContainerIcon.o" "objects/haiku/x86_64/release/kits/tracker/DialogPane.o" "objects/haiku/x86_64/release/kits/tracker/DirMenu.o" "objects/haiku/x86_64/release/kits/tracker/EntryIterator.o" "objects/haiku/x86_64/release/kits/tracker/FBCPadding.o" "objects/haiku/x86_64/release/kits/tracker/FSClipboard.o" "objects/haiku/x86_64/release/kits/tracker/FSUndoRedo.o" "objects/haiku/x86_64/release/kits/tracker/FSUtils.o" "objects/haiku/x86_64/release/kits/tracker/FavoritesMenu.o" "objects/haiku/x86_64/release/kits/tracker/FilePanel.o" "objects/haiku/x86_64/release/kits/tracker/FilePanelPriv.o" "objects/haiku/x86_64/release/kits/tracker/FilePermissionsView.o" "objects/haiku/x86_64/release/kits/tracker/FindPanel.o" "objects/haiku/x86_64/release/kits/tracker/GeneralInfoView.o" "objects/haiku/x86_64/release/kits/tracker/GroupedMenu.o" "objects/haiku/x86_64/release/kits/tracker/HeaderView.o" "objects/haiku/x86_64/release/kits/tracker/IconCache.o" "objects/haiku/x86_64/release/kits/tracker/IconMenuItem.o" "objects/haiku/x86_64/release/kits/tracker/InfoWindow.o" "objects/haiku/x86_64/release/kits/tracker/MimeTypeList.o" "objects/haiku/x86_64/release/kits/tracker/MiniMenuField.o" "objects/haiku/x86_64/release/kits/tracker/Model.o" "objects/haiku/x86_64/release/kits/tracker/MountMenu.o" "objects/haiku/x86_64/release/kits/tracker/Navigator.o" "objects/haiku/x86_64/release/kits/tracker/NavMenu.o" "objects/haiku/x86_64/release/kits/tracker/NodePreloader.o" "objects/haiku/x86_64/release/kits/tracker/NodeWalker.o" "objects/haiku/x86_64/release/kits/tracker/OpenWithWindow.o" "objects/haiku/x86_64/release/kits/tracker/OverrideAlert.o" "objects/haiku/x86_64/release/kits/tracker/PendingNodeMonitorCache.o" "objects/haiku/x86_64/release/kits/tracker/Pose.o" "objects/haiku/x86_64/release/kits/tracker/PoseList.o" "objects/haiku/x86_64/release/kits/tracker/PoseView.o" "objects/haiku/x86_64/release/kits/tracker/PoseViewScripting.o" "objects/haiku/x86_64/release/kits/tracker/QueryContainerWindow.o" "objects/haiku/x86_64/release/kits/tracker/QueryPoseView.o" "objects/haiku/x86_64/release/kits/tracker/RecentItems.o" "objects/haiku/x86_64/release/kits/tracker/RegExp.o" "objects/haiku/x86_64/release/kits/tracker/SelectionWindow.o" "objects/haiku/x86_64/release/kits/tracker/Settings.o" "objects/haiku/x86_64/release/kits/tracker/SettingsViews.o" "objects/haiku/x86_64/release/kits/tracker/SlowMenu.o" "objects/haiku/x86_64/release/kits/tracker/StatusWindow.o" "objects/haiku/x86_64/release/kits/tracker/TaskLoop.o" "objects/haiku/x86_64/release/kits/tracker/TemplatesMenu.o" "objects/haiku/x86_64/release/kits/tracker/Tests.o" "objects/haiku/x86_64/release/kits/tracker/TextWidget.o" "objects/haiku/x86_64/release/kits/tracker/Thumbnails.o" "objects/haiku/x86_64/release/kits/tracker/TitleView.o" "objects/haiku/x86_64/release/kits/tracker/Tracker.o" "objects/haiku/x86_64/release/kits/tracker/TrackerInitialState.o" "objects/haiku/x86_64/release/kits/tracker/TrackerScripting.o" "objects/haiku/x86_64/release/kits/tracker/TrackerSettings.o" "objects/haiku/x86_64/release/kits/tracker/TrackerSettingsWindow.o" "objects/haiku/x86_64/release/kits/tracker/TrackerString.o" "objects/haiku/x86_64/release/kits/tracker/TrashWatcher.o" "objects/haiku/x86_64/release/kits/tracker/Utilities.o" "objects/haiku/x86_64/release/kits/tracker/ViewState.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryEntryList.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryManager.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryPoseView.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryWindow.o" "objects/haiku/x86_64/release/kits/tracker/WidgetAttributeText.o" \
"objects/haiku/x86_64/release/kits/libbe.so" "objects/haiku/x86_64/release/kits/locale/liblocalestub.a" "objects/haiku/x86_64/release/kits/shared/libshared.a" "objects/haiku/x86_64/release/kits/translation/libtranslation.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libstdc++.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libstdc++.so" "objects/haiku/x86_64/release/kits/interface/libcolumnlistview.a" "objects/haiku/x86_64/release/add-ons/input_server/filters/shortcut_catcher/libshortcuts_shared.a" "objects/haiku/x86_64/release/system/libroot/libroot.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libgcc_s.so.1" "build_packages/gcc_syslibs_devel-13.2.0_2023_08_10-3-x86_64/develop/lib/libgcc.a"  "/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/crtendS.o" "objects/haiku/x86_64/release/system/glue/arch/x86_64/crtn.o" \


...failed Link objects/haiku/x86_64/release/kits/tracker/libtracker.so ...

BUILD FAILURE:
...failed updating 1 target(s)...
...skipped 1 target(s)...
```

Calisto-Mathias commented:
Oh I must have not pushed the Jamfile. I had added the SaveWindow.cpp file
to the list.

On Sun, 2 Jun, 2024, 3:28 pm Niels Sascha Reedijk, ***@***.***>
wrote:

> The current code seems to end up in a linker error for me.
>
> /DevData1/haiku-build> jam -q Tracker
> warning: Invalid jamfile cache: Failed to read file info.
> Starting build of type regular ...
> AddHaikuImagePackages: package gutenprint8 not available!
> Asked for bios_ia32 target boot platform
> Unknown path to handle adding to image
> Asked for pxe_ia32 target boot platform
> Unknown path to handle adding to image
> ...patience...
> ...patience...
> ...found 13655 target(s)...
> ...updating 2 target(s)...
> Link objects/haiku/x86_64/release/kits/tracker/libtracker.so
> /boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: objects/haiku/x86_64/release/kits/tracker/FindPanel.o: in function `BPrivate::FindWindow::MessageReceived(BMessage*)':
> FindPanel.cpp:(.text+0x8814): undefined reference to `BPrivate::SaveWindow::SaveWindow(BWindow*)'
> collect2: error: ld returned 1 exit status
>
> gcc -fno-strict-aliasing -fno-delete-null-pointer-checks -fno-builtin-fork -fno-builtin-vfork -Xlinker --no-undefined -shared -Xlinker -soname="libtracker.so" -nostdlib -Xlinker --no-undefined -o "objects/haiku/x86_64/release/kits/tracker/libtracker.so"  "objects/haiku/x86_64/release/system/glue/arch/x86_64/crti.o" "/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/crtbeginS.o" "objects/haiku/x86_64/release/system/glue/init_term_dyn.o" "objects/haiku/x86_64/release/kits/tracker/AttributeStream.o" "objects/haiku/x86_64/release/kits/tracker/AttributesView.o" "objects/haiku/x86_64/release/kits/tracker/AutoMounterSettings.o" "objects/haiku/x86_64/release/kits/tracker/BackgroundImage.o" "objects/haiku/x86_64/release/kits/tracker/Bitmaps.o" "objects/haiku/x86_64/release/kits/tracker/ContainerWindow.o" "objects/haiku/x86_64/release/kits/tracker/CountView.o" "objects/haiku/x86_64/release/kits/tracker/DeskWindow.o" "objects/haiku/x86_64/release/kits/tracker/DesktopPoseView.o" "objects/haiku/x86_64/release/kits/tracker/DraggableContainerIcon.o" "objects/haiku/x86_64/release/kits/tracker/DialogPane.o" "objects/haiku/x86_64/release/kits/tracker/DirMenu.o" "objects/haiku/x86_64/release/kits/tracker/EntryIterator.o" "objects/haiku/x86_64/release/kits/tracker/FBCPadding.o" "objects/haiku/x86_64/release/kits/tracker/FSClipboard.o" "objects/haiku/x86_64/release/kits/tracker/FSUndoRedo.o" "objects/haiku/x86_64/release/kits/tracker/FSUtils.o" "objects/haiku/x86_64/release/kits/tracker/FavoritesMenu.o" "objects/haiku/x86_64/release/kits/tracker/FilePanel.o" "objects/haiku/x86_64/release/kits/tracker/FilePanelPriv.o" "objects/haiku/x86_64/release/kits/tracker/FilePermissionsView.o" "objects/haiku/x86_64/release/kits/tracker/FindPanel.o" "objects/haiku/x86_64/release/kits/tracker/GeneralInfoView.o" "objects/haiku/x86_64/release/kits/tracker/GroupedMenu.o" "objects/haiku/x86_64/release/kits/tracker/HeaderView.o" "objects/haiku/x86_64/release/kits/tracker/IconCache.o" "objects/haiku/x86_64/release/kits/tracker/IconMenuItem.o" "objects/haiku/x86_64/release/kits/tracker/InfoWindow.o" "objects/haiku/x86_64/release/kits/tracker/MimeTypeList.o" "objects/haiku/x86_64/release/kits/tracker/MiniMenuField.o" "objects/haiku/x86_64/release/kits/tracker/Model.o" "objects/haiku/x86_64/release/kits/tracker/MountMenu.o" "objects/haiku/x86_64/release/kits/tracker/Navigator.o" "objects/haiku/x86_64/release/kits/tracker/NavMenu.o" "objects/haiku/x86_64/release/kits/tracker/NodePreloader.o" "objects/haiku/x86_64/release/kits/tracker/NodeWalker.o" "objects/haiku/x86_64/release/kits/tracker/OpenWithWindow.o" "objects/haiku/x86_64/release/kits/tracker/OverrideAlert.o" "objects/haiku/x86_64/release/kits/tracker/PendingNodeMonitorCache.o" "objects/haiku/x86_64/release/kits/tracker/Pose.o" "objects/haiku/x86_64/release/kits/tracker/PoseList.o" "objects/haiku/x86_64/release/kits/tracker/PoseView.o" "objects/haiku/x86_64/release/kits/tracker/PoseViewScripting.o" "objects/haiku/x86_64/release/kits/tracker/QueryContainerWindow.o" "objects/haiku/x86_64/release/kits/tracker/QueryPoseView.o" "objects/haiku/x86_64/release/kits/tracker/RecentItems.o" "objects/haiku/x86_64/release/kits/tracker/RegExp.o" "objects/haiku/x86_64/release/kits/tracker/SelectionWindow.o" "objects/haiku/x86_64/release/kits/tracker/Settings.o" "objects/haiku/x86_64/release/kits/tracker/SettingsViews.o" "objects/haiku/x86_64/release/kits/tracker/SlowMenu.o" "objects/haiku/x86_64/release/kits/tracker/StatusWindow.o" "objects/haiku/x86_64/release/kits/tracker/TaskLoop.o" "objects/haiku/x86_64/release/kits/tracker/TemplatesMenu.o" "objects/haiku/x86_64/release/kits/tracker/Tests.o" "objects/haiku/x86_64/release/kits/tracker/TextWidget.o" "objects/haiku/x86_64/release/kits/tracker/Thumbnails.o" "objects/haiku/x86_64/release/kits/tracker/TitleView.o" "objects/haiku/x86_64/release/kits/tracker/Tracker.o" "objects/haiku/x86_64/release/kits/tracker/TrackerInitialState.o" "objects/haiku/x86_64/release/kits/tracker/TrackerScripting.o" "objects/haiku/x86_64/release/kits/tracker/TrackerSettings.o" "objects/haiku/x86_64/release/kits/tracker/TrackerSettingsWindow.o" "objects/haiku/x86_64/release/kits/tracker/TrackerString.o" "objects/haiku/x86_64/release/kits/tracker/TrashWatcher.o" "objects/haiku/x86_64/release/kits/tracker/Utilities.o" "objects/haiku/x86_64/release/kits/tracker/ViewState.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryEntryList.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryManager.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryPoseView.o" "objects/haiku/x86_64/release/kits/tracker/VirtualDirectoryWindow.o" "objects/haiku/x86_64/release/kits/tracker/WidgetAttributeText.o" \
> "objects/haiku/x86_64/release/kits/libbe.so" "objects/haiku/x86_64/release/kits/locale/liblocalestub.a" "objects/haiku/x86_64/release/kits/shared/libshared.a" "objects/haiku/x86_64/release/kits/translation/libtranslation.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libstdc++.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libstdc++.so" "objects/haiku/x86_64/release/kits/interface/libcolumnlistview.a" "objects/haiku/x86_64/release/add-ons/input_server/filters/shortcut_catcher/libshortcuts_shared.a" "objects/haiku/x86_64/release/system/libroot/libroot.so" "build_packages/gcc_syslibs-13.2.0_2023_08_10-3-x86_64/lib/libgcc_s.so.1" "build_packages/gcc_syslibs_devel-13.2.0_2023_08_10-3-x86_64/develop/lib/libgcc.a"  "/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/crtendS.o" "objects/haiku/x86_64/release/system/glue/arch/x86_64/crtn.o" \
>
>
> ...failed Link objects/haiku/x86_64/release/kits/tracker/libtracker.so ...
>
> BUILD FAILURE:
> ...failed updating 1 target(s)...
> ...skipped 1 target(s)...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Calisto-Mathias/haiku/pull/1#issuecomment-2143780889>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APE5B6EXV272LSMAUDCGGDDZFLUFLAVCNFSM6AAAAABIQ7UB4KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNBTG44DAOBYHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Calisto-Mathias commented:
The recent commit includes a new Jamfile that integrates the compilation of the SaveWindow.cpp file. Additionally, it consolidates almost all functionalities of the current options menu and introduces the feature of opening a query file directly into the Find Panel via the menu bar.

In the current version of Tracker, saving a query to a preferred destination is not possible. This new functionality could be beneficial in several scenarios. The updated menu bar now supports saving a query, either as a regular query or as a template, in the user's default directory or a chosen location. It also includes a separate section in the menu bar for listing favorites.

So, what differentiates this from the regular feature that allows saving a query as a template?

The primary difference lies in how each is opened. Clicking on a template opens it in the Find panel, allowing you to edit the fields and other details before executing the query. Upon execution, a query with an appropriate name is created in the user's default directory. In contrast, double-clicking on a query directly displays its results in the QueryPoseView format.

This new functionality may only minimally reduce execution time, but it was implemented as part of moving the functionality into a menu bar. This reorganization streamlines the user interface and enhances accessibility, providing users with a slightly more intuitive and efficient experience.

nielx commented:
Will do a more full review later, but I have tried out the changes and I can see where this is going!
Based on our conversation, and me seeing this work, I have the following thoughts/questions;

- I like the menu-based find window more than I thought I would, so it is worth continuing in this path.
- About the icon: as we discussed during the call, is it possible to position and size it like the main tracker window? I.e. position it in the menu bar.
- What are you planning on putting in the attributes menu?
- I saw the 'Open Query' menu item lead to a crash. I will look at the code to see why that is.
- What is the purpose of the 'formula' menu item?
- And does the 'Temporary Query' menu item have a meaning if all queries are temporary unless saved? 
- Likewise: from that thinking: what does that mean to 'Open' a query. If I open it, am i then editing it? And how does the saving work then? Will it overwrite, like a normal document?

Calisto-Mathias commented:
I'm glad to hear that you find the menu-bar-based find window better than the previous version. Could you please email me the crash report? The window doesn't crash on my system and it opens up a Find Panel, but there might be something else wrong with that section of the code. I will take a look into it to identify the cause of the crash.

In addition, I have temporarily removed the Edit Formula and Attributes Menu options, as they don't seem to make much sense at the moment since we haven't transitioned to the new layout (and we might not, eventually). Initially, I thought the current "Find By Attribute" section of the find panel, which only allows searching by Name, Size, and Modified, could be expanded. By letting users select more attributes, the search experience could be significantly improved.

For example, with the Person file type, a user might want to find all the people from a specific company to email them about an issue. Such a query wouldn't be possible through the current "Find By Attribute" menu. My idea is that, after implementing incremental search, I could add the ability to search by individual attributes as well. This would greatly enhance the search capabilities of the find window.

For now, I have removed these options from the menu as they were empty. I will revisit this after the incremental search feature is implemented.

humdingerb commented:
> Likewise: from that thinking: what does that mean to 'Open' a query. If I open it, am i then editing it? And how does the saving work then? Will it overwrite, like a normal document?

This is worth thinking about. Also: what's a "recent query"? We originally planned to keep a list of the 10 most recent queries. But with the incremental/live search this needs to change.
We need to distinguish between
* "Queries" (= filetype + attributes + column layout + search terms)
* "Query template" (= filetype + attributes + column layout).

So it makes only sense to have recent "Query templates" in a list, or it'll grow with every keystroke...

The "Query template" of every use of the find panel will continue to be saved to ~/queries where it self-destructs after 7 days if it's marked "temporary" (not sure that's actually working ATM, as I have some quite old queries in ~/queries).

When saving a query, the users decide if they want to save as Query (incl. search terms) or Template (no search terms).

Calisto-Mathias commented:
Hi Niels,

I've successfully located the code responsible for query execution. I'm now working on finding a suitable way to reference these windows. Currently, I found a function that performs a straightforward task: the Find Panel closes the old search query window and opens a new one if the query being edited and searched already has an open window. It does this using the EntryHasWindowOpen(entry_ref*) function, which locks the list of open windows maintained by Tracker. It then calls the FindContainerWindow function with the same entry_ref. This FindContainerWindow function returns a BContainerWindow*, which I could use to reference the window. Through this window, I can access the underlying PoseView using the public PoseView getter function.

What I'm thinking of right now is to add some code here to edit this function or create a new helper function that returns a BMessenger targeting the PoseView. Using this, I can then send messages to the Query Pose View and implement the incremental search by looping through all the entries of the query and editing the query as needed. This could potentially eliminate the need to create a file every time a search query is executed, as Humdinger pointed out. Instead of having the Tracker receive a B_REFS_RECEIVED message that instructs it to open the file, I could directly run the OpenRef function in the Find Panel, which would open the results window. Subsequently, I could receive the BMessenger, which would remain open as long as the Find Panel is open and automatically close both the results window and the Find Panel upon receiving a quit message.

Additionally, I could modify the Query results window to prevent it from being closed while the Find Panel is open. This approach aligns with opening the Find Panel whenever a query or query template is selected. Currently, opening a query template loads it in the Find Panel, but opening a query directly loads the results window. This behavior wouldn't make sense with incremental search. Instead, clicking on the option to open a query should open both the Find Panel and a results window, which would automatically execute the search upon opening the Find Panel. The incremental search functionality would remain the same.

The query fetching code is part of the QueryPoseView.cpp file and runs during the instantiation of a QueryEntryListCollection. Instead of sending the pointer and allowing direct access, I was thinking of writing a small function that returns a BMessenger to the Query Pose Window. This BMessenger could then be returned to the Find Panel and used for any subsequent changes via a private instance variable. Does this approach make sense?

If so, I'll get started on the task.

nielx commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -253,7 +256,20 @@ FindWindow::FindWindow(const entry_ref* newRef, bool editIfTemplateOnly)
 		fEditTemplateOnly);
-	SetLayout(new BGroupLayout(B_VERTICAL));
+		
+	fMenuBar = new BMenuBar("Menu Bar");
```
Why doesn't the BuildMenuBar() function not create the main menu bar?

Calisto-Mathias replied:
Fixed this in the latest commit.

nielx commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -650,6 +768,76 @@ FindWindow::MessageReceived(BMessage* message)
+
+		case kCloseSaveQueryPanel:
+		{
+			const char *queryName;
```
The * should be to the left, it is part of the type.

There are multiple occurrences in your code.

Calisto-Mathias replied:
Oh. I fixed this throughout the files. I will definitely keep this in mind.

nielx commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -650,6 +768,76 @@ FindWindow::MessageReceived(BMessage* message)
+
+			if(message->FindString("Query Name", &queryName) == B_OK
+				&& message->FindBool("Include In Favorites", &includeInFavorites) == B_OK
+				&& message->FindBool("Save In Default Directory", &saveInDefaultDirectory )== B_OK){
```
space between ) and {

This is in multiple files.

Calisto-Mathias replied:
Fixed in the Latest commit.

nielx commented on src/kits/tracker/SaveWindow.cpp:
```diff
@@ -0,0 +1,133 @@
+/*
```
License: I think this code is mostly/all completely new, so you don't use the old OpenTracker license, but the modern Haiku license. Check the coding guidelines for more information.

Also in the header file.

nielx commented:
With the notes you shared, I have re-reviewed the PR and I have two remaining points:

- The spacing of the Save Window seems off. There is a lot of space between the checkboxes, and the Save button itself seems to be in a non-standard position. If you can post a screenshot, @humdingerb could probably advise.
- Given this new functionality, isn't any query that is not saved, a Temporary Query? I.e. what is the purpose of that menu option?

Suggestions for subsequent improvements (i.e. not in this PR):

- I think it is worth exploring if we can get rid of the save window. It seems to have a overlap with the save panel (i.e. the name field). I would try customizing the BFilePanel and add buttons/check boxes to copy the default location and favorite functionality.
- If we are going full document mode on queries, then a normal 'Save' action to update queries that have been previously saved, is a logical addition.

Let's get this PR wrapped up! Did you share the draft for the forum post with Humdinger as well?

nielx commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -262,7 +282,222 @@ FindWindow::FindWindow(const entry_ref* newRef, bool editIfTemplateOnly)
+	// Simple Helper Function To Check For Duplicates Within an Entry List of Templates
+	int32 count = list->CountItems();
+	for(int32 i = 0;i<count;i++) {
+		if(strcmp(BPath(list->ItemAt(i)).Path(), BPath(ref).Path()) == 0) {
```
BPath has implemented the == operator (see https://www.haiku-os.org/docs/api/classBPath.html#a3df539180ce332bebd6aa5b657b76254)

Calisto-Mathias replied:
Yup. I fixed this in the latest commit.

nielx commented on src/kits/tracker/FindPanel.cpp:
```diff
@@ -262,7 +282,222 @@ FindWindow::FindWindow(const entry_ref* newRef, bool editIfTemplateOnly)
+
+				char type[B_MIME_TYPE_LENGTH];
+				BNodeInfo(new BNode(&ref)).GetType(type);
+				if (strcmp(type, B_QUERY_TEMPLATE_MIMETYPE) == 0 && BEntry(&ref).Exists()
```
I would add a comment why you are checking for duplicates

Calisto-Mathias replied:
I added a small comment below the line.

nielx reviewed (commented):
See comments
